System/Metrics.hs

```diff
@@ -114,109 +155,308 @@ import qualified System.Metrics.Label as Label
+  {-# INLINE toTags #-}
 
-type GroupId = Int
+-- | Disallow tags altogether.
+--
+-- > toTags () = HashMap.empty
+instance ToTags () where
+  toTags () = M.empty
+  {-# INLINE toTags #-}
 
--- | The 'Store' state.
-data State = State
-     { stateMetrics :: !(M.HashMap T.Text (Either MetricSampler GroupId))
-     , stateGroups  :: !(IM.IntMap GroupSampler)
-     , stateNextId  :: {-# UNPACK #-} !Int
-     }
+-- | Place no constraints on tags.
+--
+-- > toTags @(HashMap Text Text) = id
+instance ToTags (M.HashMap T.Text T.Text) where
+  toTags = id
+  {-# INLINE toTags #-}
 
-data GroupSampler = forall a. GroupSampler
-     { groupSampleAction   :: !(IO a)
-     , groupSamplerMetrics :: !(M.HashMap T.Text (a -> Value))
-     }
+------------------------------------------------------------------------
+-- ** Deriving `ToTags`
+--
+-- | Deriving instances of `ToTags` for records that exclusively have
+-- fields of type `Text`.
+class GToTags (f :: Type -> Type) where
+  gToTags :: T.Text -> f x -> M.HashMap T.Text T.Text
+
+-- Data (passthrough)
+instance (GToTags f) => GToTags (D1 c f) where
+  gToTags name (M1 x) = gToTags name x
+  {-# INLINE gToTags #-}
+
+-- Constructor (passthrough)
+instance (GToTags f) => GToTags (C1 c f) where
+  gToTags name (M1 x) = gToTags name x
+  {-# INLINE gToTags #-}
+
+-- Products (union)
+instance (GToTags f, GToTags g) => GToTags (f :*: g) where
+  gToTags name (x :*: y) = gToTags name x `M.union` gToTags name y
+  {-# INLINE gToTags #-}
+
+-- Record selectors (take record selector name)
+instance (GToTags f, KnownSymbol name) =>
+  GToTags (S1 ('MetaSel ('Just name) su ss ds) f) where
+  gToTags _name (M1 x) =
+    let name' = T.pack $ symbolVal $ Proxy @name
+    in  gToTags name' x
+  {-# INLINE gToTags #-}
+
+-- Individual fields (take value, combine with name)
+instance GToTags (K1 i T.Text) where
+  gToTags name (K1 x) = M.singleton name x
+  {-# INLINE gToTags #-}
 
--- TODO: Rename this to Metric and Metric to SampledMetric.
-data MetricSampler = CounterS !(IO Int64)
-                   | GaugeS !(IO Int64)
-                   | LabelS !(IO T.Text)
-                   | DistributionS !(IO Distribution.Stats)
+------------------------------------------------------------------------
+-- * Changing scope
 
--- | Create a new, empty metric store.
-newStore :: IO Store
-newStore = do
-    state <- newIORef $ State M.empty IM.empty 0
-    return $ Store state
+-- $scopes
+-- References to metric stores are parameterized by types that restrict
+-- the kinds of metrics that may be added to or removed from the store
+-- through the references. In other words, they are /scoped/ by their
+-- type parameter.
+--
+-- It can be useful to create a new reference to a metric store that is
+-- scoped to a subset of its metrics. This can be done as long as the
+-- subset can be represented by a function (see `subset`).
+
+-- | Create a new reference to a metric store with restricted scope.
```

Comment:
Might be worth adding an example in the published comments along with a justification (in a non-published comment) for why `coerce` is a safe implementation.

Reply:
Sounds good!

Though, I suppose this function is not safe because it never evaluates its first argument. This allows a user to change scope arbitrarily via `subset undefined`. I don't know how to prevent the use of `undefined` ... I was hoping that we could trust users to not use `undefined`, or maybe allow them to use it as an escape hatch.

Reply:
Thanks for adding the example - can you please explain the intended purpose of the scopes/subset functionality though, because I'm not seeing it.

Reply:
Sure! I had two motivations when introducing these "scopes"/"subsets".

(1) I wanted to preserve an `ekg-core` feature in the presence of the type-annotated stores introduced in this fork. The `ekg-core` feature is that users can register to their store metrics defined by other code (i.e. libraries). Since this fork requires that all metrics be statically defined, the library code must register their metrics under some metrics specification. "Subsets" were a convenient way for users to include those library-defined metrics specifications in their own metrics specifications, thereby allowing users to register library-defined metrics. I didn't think further than this, though. Perhaps there is another solution not involving these "scopes".

In my integration of this fork with our codebase, I use "scopes" to register the pre-defined RTS metrics provided by `ekg-core`.

(2) `ekg-core` supports only static metrics, but this fork also supports dynamic metrics (that is, metrics can be deregistered and re-registered). This support for dynamic metrics expands the number of programs that compile and run without error. For example, in `ekg-core`, registering a metric that already exists results in an exception, whereas in this fork the existing metric would be silently replaced. Using "scopes" was a way to recover static metrics (i.e. those metrics that don't need to be -- and shouldn't be -- deregistered). To make static metrics with scopes, you would (a) separate your dynamic and static metrics into two "subsets" of your metrics specification, (b) register the static metrics at startup, and (c) create a new reference to the store scoped only to the dynamic metrics and discard the original store reference. This statically guarantees that the metrics in the "static" subset can no longer be deregistered.

In my integration of this fork with our codebase, I used this "technique", but it is not essential.

Reply:
I've added source code comments explaining why `subset`'s implementation is safe, and added a section to the tutorial showing how to simulate static metrics (what I tried to explain in the above part (b)).

Although the scopes/subsets may be a sub-optimal API and may need to be revised or replaced, we will have the opportunity to do so because this library will only be used internally for some time. Then, I'm going to go ahead and merge the current API with scopes/subsets with the intention of returning to it later.